plugins_src/import_export/wpc_jscad.erl

```diff
@@ -0,0 +1,189 @@
+    Exporter(Ps, export_fun(Attr)).
+
+export_fun(Attr) ->
+    fun(Filename, Contents) ->
+	export(Filename, Contents, Attr)
+    end.
+
+export(Filename, Contents0, Attr) ->
+    Contents = export_transform(Contents0, Attr),
+    #e3d_file{objs=Objs,mat=Mat,creator=Creator} = Contents,
+    {ok,F} = file:open(Filename, [write]),
+    FName = filename:basename(Filename)++filename:extension(Filename),
+    NObjs = length(Objs),
+    %% Write file head
+    io:format(F, "// File         : ~ts\n",[FName]),
+    io:format(F, "// Objects      : ~lp\n",[NObjs]),
+    io:format(F, "// Exported from: ~ts\n",[Creator]),
+    %% Write file body
+    io:format(F, "function main() {\n",[]),
+    io:format(F, "  return [\n",[]),
+    try
+	lists:foldl(fun(#e3d_object{name = Name, obj=Obj}, AccObj) ->
+			Meshes = e3d_mesh:split_by_material(Obj),
+			NMeshes = length(Meshes),
+			all(fun(Mesh,AccMeshes) ->
+				export_object(F, Name, Mesh, Mat),
+				AccMeshes-1
+			    end, F, NMeshes, Meshes),
+			if AccObj > 1 ->
+			    io:put_chars(F, ",\n");
+			    true -> ignore
+			end,
+			AccObj-1
+		  end, NObjs, Objs),
+    	io:format(F, "\n  ];\n}\n",[])
+    catch _:Err:Stacktrace ->
+	io:format(?__(1,"OpenJSCAD Error: ~P in")++" ~p~n", [Err,30,Stacktrace])
+    end,
+    ok = file:close(F).
+
+export_object(F, ObjName, #e3d_mesh{type=Type,fs=Fs,vs=VTab}, Mat_defs) ->
+    [#e3d_face{mat=[Material|_]}|_] = Fs,
+    DifColor = material(Material, Mat_defs),
+    io:format(F, "    // ~ts.~ts\n",[ObjName,Material]),
+    io:format(F, "    color(~p, \n",[DifColor]),
+    io:format(F, "      polyhedron(\n",[]),
+
+    %% Helpers
+    W3 = fun({X,Y,Z}) ->  io:format(F, "                    [~p,~p,~p]", [X,Y,Z]) end,
+    %% Write vertex coordinates
+    io:format(F, "        { points: [\n",[]),
+    all(W3,F,VTab),
+    io:format(F, "\n                  ],\n",[]),
+    %% Write vertex indexes for faces
+    SType =
+	case Type of
+	    triangle -> "triangles";
+	    _ -> "polygons"
+	end,
+    io:format(F, "          ~s: [\n",[SType]),
+    Ident = "          "++string:right("",length(SType))++"     ",
+    all(fun(#e3d_face{vs=Vs}) -> print_face(F,Ident,Vs) end,F,Fs),
+    io:format(F, "\n          ~s  ]\n",[string:right("",length(SType))]),
+    %% Close polyhedron's data (points and triangles/polygons)
+    io:put_chars(F, "        }\n"),
+    %% Close color and polyhedron
+    io:put_chars(F, "      )\n"),
+    io:put_chars(F, "    )").
+
+material(Name, Mat_defs) ->
+    MatInfo = lookup(Name, Mat_defs),
+    Mat = lookup(opengl, MatInfo),
+    {Dr, Dg, Db, _} = lookup(diffuse, Mat),
+    [Dr, Dg, Db].
+
+print_face(F, Ident, Vs0) ->
+    Vs = Vs0,
+    io:put_chars(F, Ident++"["),
+    %% we need to invert the face order to get it in CW sequence as OpenJSCAD needs
+    lists:foldr(fun(V,1) ->
+		    io:format(F, "~lp", [V]);
+	       (V,Acc) ->
+		    io:format(F, "~lp,", [V]),
+		   Acc-1
+	       end, length(Vs), Vs),
```

Comment:
Doesn't  `io:format(F,"~w", [lists:reverse(V)]),` work here?

Reply:
All this time and I didn't realize this char format would be applied to a entire list. 🤦‍♂